Name: __________________        Class: __________________     Date: __________________

<center>
    <img width="100%" src="https://raw.githubusercontent.com/astroDimitrios/AstronomyClass/master/DesignAssets/Ai/Banner3plainCut.png" alt='AP Logo'>
</center>

# Spreading Out

### AIM - To perform simple calculations using Python
Difficulty: Easy

## Start:

This activity is designed to help you improve your basic Python and introduce you to string formatting and function arguments. You are going to calculate how much power is generated by the solar panels of the Mars Reconnaisance Orbiter ([MRO](https://mars.nasa.gov/mro/)). 

Activity inspired by a homework activity by Prof. A. Spitkovsky [here](https://www.astro.princeton.edu/~anatoly/ast203/Homework_3_soln).

## Contents

* [Start](#Start)
* [Terminology](#Term)
* [Inverse Square Laws](#INV)
* [Iterating with enumerate](#ENU)
* [String Formatting](#STRING)
* [NumPy Where](#WHERE)
* [Function Arguments](#FA)
* [Docstrings](#DS)
* [Over to You](#OTY)


## Let's Go: <a class="anchor" id="Start"></a>

The first thing you should do is import the libraries you think you'll need like NumPy. I suggest you start by importing NumPy below:

In [1]:
# Don't forget the as np
import numpy as np

You are going to start by calculating the Luminosity of the Sun. To do this we need some constants which I have defined for you in the code cell below.

* ```sigma``` - $\sigma$ is the Stefan-Boltzmann constant
* ```Rsun``` is the radius of the Sun
* ```Tsun``` is the temperature of the Sun's surface
* ```AU``` is the conversion from astronomical units to metres

Run the cell below to declare the constants.

In [2]:
sigma = 5.670373*10**(-8) # Wm^-2K^-4
Rsun = 696340000 # m
Tsun = 5778 # K
AU = 1.495978707*10**11 # m

### Terminology <a class="anchor" id="Term"></a>

A quick note on terminology. I have used three terms throughout this notebook, **Power**, **Luminosity**, and **Brightness**. Before you move on I've defined these terms for clarity.
* **Power** is the energy transferred per second ($J/s = W$, Watts)
* **Luminosity** is the power of a light emitting object ($J/s = W$, Watts)
* **Brightness** is how bright the object looks from a distance and is measured in Watts per metre squared ($Wm^{-2}$)

So in this notebook when we are talking about the luminosity of the Sun you can think of it as the power of the Sun.

To calculate the Luminosity of the Sun we can use the [Stefan-Boltzmann Law](https://en.wikipedia.org/wiki/Stefan%E2%80%93Boltzmann_law) for a perfect blackbody:

$$L = A\sigma T^4$$ 

where $A$ is the surface area of a sphere given by:

$$A = 4\pi r^2$$

In the code cell below use the Stefan-Boltzmann Law to calculate the Luminosity of the Sun and store it in the variable ```Lsun```. Luminosity is similar to Power. Both are measured in Watts, which is the energy transferred per second in J/s. Print the variable.

In [3]:
Lsun = 4*np.pi*Rsun**2*sigma*Tsun**4
Lsun

3.8510071797476947e+26

You should have calculated $3.85\times 10^{26}\ W$! That's a huge amount of energy transferred per second from the Sun.

To calculate how much power we can use at Mars we will need a list of planet names and a list of distances. These are provided for you below. Run the code cell below to declare the lists.

In [4]:
names = ['Mercury', 'Venus', 'Earth', 'Mars', 'Jupiter', 'Saturn', 'Neptune', 'Uranus', 'Pluto']
distances = [0.4, 0.7, 1.0, 1.5, 5.2, 9.6, 30, 19.2, 39.5] # These are in astronomical units (AU)

It is often easier to work with data stored in NumPy arrays rather than lists. We can convert them to NumPy arrays using the ```np.asarray()``` command. The first conversion has been done for you. Convert the distances list to a NumPy array.

In [5]:
names = np.asarray(names)
distances = np.asarray(distances)

The Sun is a giant glowing ball of gas. Its energy is radiated outward as light in all directions almost uniformly. 
The power that Mars gets from the Sun depends on two factors. The Luminosity of the Sun, and the distance Mars is from the Sun.

The brightness drops off with the distance squared. The brightness $b$ is the power you get per metre squared - $Wm^{-2}$. The formula for brightness is an example of an inverse square law which take the form:

$$intensity \propto \frac{1}{distance^2}$$ <a class="anchor" id="INV"></a>

Here if we double the distance of our planet from the Sun the brightness drops by a factor of $2^2$ or $4$.    
The formula for brightness is:

$$b = \frac{L}{4\pi r^2}$$

where $b$ is the brightness, $L$ is the luminosity of the object radiating light, and $r$ is how far from the object you are. The denominator is just the surface area of a sphere with radius $r$ since the light from the object radiates spherically outwards like in this figure:

<img align="center" src="https://raw.githubusercontent.com/astroDimitrios/astroDimitrios.github.io/master/images/InverseSq2.jpg" alt="Inverse Square Image" align="center">

The spheres have been cut in half here so that you can see the inner spheres!

Use the code block below to calculate the brightness for all the distances. You can do this by performing the calculation on the whole array not just one element at a time. Store your array of brightness values in a variable named ```b``` and print it. Don't forget the $L$ in the equation is the ```Lsun``` variable we already calculated and convert from AU to m! 

In [6]:
b = Lsun/(4*np.pi*(distances*AU)**2) # W/m^2
b

array([8.55842122e+03, 2.79458652e+03, 1.36934740e+03, 6.08598843e+02,
       5.06415457e+01, 1.48583702e+01, 1.52149711e+00, 3.71459255e+00,
       8.77646144e-01])

Now you have an array of the brightnesses (power per metre squared remember) at each of the planets orbits.

The MRO has two solar panels with a combined area of $10\ m^2$ and an efficiency of $0.3$. To calculate the power from the solar panels all you need to do is times the brightness array ```b``` by the ```area``` and ```efficiency``` constants which have been defined for you below.

Run the cell below with the constants, then in the cell below that calculate the power from the solar panels. Store the array in a variable called ```Pmro``` and print it.

In [7]:
area = 10 # m^2
efficiency = .3

In [8]:
Pmro = b*area*efficiency
Pmro

array([2.56752637e+04, 8.38375957e+03, 4.10804219e+03, 1.82579653e+03,
       1.51924637e+02, 4.45751105e+01, 4.56449132e+00, 1.11437776e+01,
       2.63293843e+00])

Yay! You have calculated the power from the solar panels of the orbiter around different planets.

The fourth value of $1.82579653\times10^3\ W$ is for Mars. Do you know intuitively how much power that is? Probably not, let's compare it to the power of a $60\ W$ lightbulb. Run the code cell below to declare our ```bulb``` constant.

In [9]:
bulb = 60 # W

To work out how many lightbulbs the solar panels could power we divide ```Pmro``` by ```bulb```. Go ahead and do that in the next code cell and then print out the variable ```Pmro_ratio```.

In [10]:
Pmro_ratio = Pmro/bulb
Pmro_ratio

array([4.27921061e+02, 1.39729326e+02, 6.84673698e+01, 3.04299421e+01,
       2.53207729e+00, 7.42918509e-01, 7.60748553e-02, 1.85729627e-01,
       4.38823072e-02])

This still isn't very readable. Let's use the ```enumerate()``` function to loop through all the planets and print out a nice sentence. <a class="anchor" id="ENU"></a>

We can use enumerate like this:

```Python
animals = ['giraffe', 'penguin', 'dolphin']

for i, animal in enumerate(animals):
    print(f'Animal {i}: {animal}')
```

This would print the following:

```
Animal 0: giraffe
Animal 1: penguin
Animal 2: dolphin
```
```animal``` gave us the actual value stored at that index.     
```i``` gave us the index of the element in the animals list. This gives us the advantage of keeping track of the index related to each animal so that we can use it later. If we tried to loop over the animals list like below we would have to create a separate variable to keep track of ```i``` and increment it every step:

```Python
index = 0
for i in animals:
    print(index, animals)
    index += 1
```
I think we can agree using enumerate is much cleaner! In our code we will need the index to access the correct values in the ```Pmro``` and ```Pmro_ratio``` arrays for that planet.    
```enumerate()``` is an example of an iterator. It allows us to cycle through all the values in our list with our for loop and often has some advantages memory-wise.

Complete the code cell below to format a string which will output the following sentence:

```
At PLANET_NAME the power is POWER W which is equivalent to POWER_RATIO lightbulbs!
```

Replace the capitalised words with ```{}``` and decide which variables go inside each ``{}```.

In [11]:
for i, name in enumerate(names):
    print(f'At {name} the power is {Pmro[i]} W which is equivalent to {Pmro_ratio[i]} lightbulbs!')

At Mercury the power is 25675.263671239303 W which is equivalent to 427.9210611873217 lightbulbs!
At Venus the power is 8383.759566118955 W which is equivalent to 139.7293261019826 lightbulbs!
At Earth the power is 4108.042187398288 W which is equivalent to 68.46736978997146 lightbulbs!
At Mars the power is 1825.7965277325725 W which is equivalent to 30.42994212887621 lightbulbs!
At Jupiter the power is 151.92463710792487 W which is equivalent to 2.532077285132081 lightbulbs!
At Saturn the power is 44.57511054034601 W which is equivalent to 0.7429185090057668 lightbulbs!
At Neptune the power is 4.564491319331431 W which is equivalent to 0.07607485532219052 lightbulbs!
At Uranus the power is 11.143777635086503 W which is equivalent to 0.1857296272514417 lightbulbs!
At Pluto the power is 2.632938431275942 W which is equivalent to 0.043882307187932366 lightbulbs!


This is better but our numbers are still too long for it to be easily readable.
<a class="anchor" id="STRING"></a>
We can format the numbers using the string formatting [mini-language](https://docs.python.org/3/library/string.html#format-specification-mini-language).
The mini-language goes inside the ```{}``` in our string after the variable name or expression. It tells Python how to format the value in the brackets.

Let's change the first bracket to ```{name:<7}```. This tells Python to left align the string within a seven character space limit and leave the spaces to the right blank.

Change the second bracket to be ```{Pmro[i]:.0f}```. This tells Python to format the power as a float with zero decimal places.

Finally change the last ```{}``` so that it formats the value as a float with one decimal place.

In [12]:
for i, name in enumerate(names):
    print(f'At {name:<7} the power is {Pmro[i]:.0f} W which is equivalent to {Pmro_ratio[i]:.1f} lightbulbs!')

At Mercury the power is 25675 W which is equivalent to 427.9 lightbulbs!
At Venus   the power is 8384 W which is equivalent to 139.7 lightbulbs!
At Earth   the power is 4108 W which is equivalent to 68.5 lightbulbs!
At Mars    the power is 1826 W which is equivalent to 30.4 lightbulbs!
At Jupiter the power is 152 W which is equivalent to 2.5 lightbulbs!
At Saturn  the power is 45 W which is equivalent to 0.7 lightbulbs!
At Neptune the power is 5 W which is equivalent to 0.1 lightbulbs!
At Uranus  the power is 11 W which is equivalent to 0.2 lightbulbs!
At Pluto   the power is 3 W which is equivalent to 0.0 lightbulbs!


This result is even better than before. There is a problem with the Pluto sentence though. Can you see it?

The last value was rounded to $0.0$! While the number is low it definetely isn't zero. Let's instead round the last value to two significant figures. To do this I have split the sentence into two parts otherwise the code would be unreadable. The first part of the sentence ```p1``` is unchanged. The second part ```p2``` now has the mini format ```:g``` which is the [general format](https://docs.python.org/3/library/string.html#format-specification-mini-language). Notice there is a string formatting call inside the outer f-string!!!

```Python
p2 = f"{float(f'{Pmro_ratio[i]:.2g}'):g} lightbulbs!"
```

The inner ```f'{Pmro_ratio[i]:.2g}'``` here rounds the number to two significant figures but formats the number in scientific notation which we don't want. To remove the scientific notation we convert this rounded string to a float using the ```float()``` function. Then we format the float again in the general format and since we aren't rounding it won't convert to scientific notation. Notice how the inner f string uses```'``` where the outer f-string uses ```"``` so python doesn't mix the two up.

Run the code cell below to see it for yourself.

In [13]:
for i, name in enumerate(names):
    p1 = f'At {name:<7} the power is {Pmro[i]:.0f} W which is equivalent to '
    p2 = f"{float(f'{Pmro_ratio[i]:.2g}'):g} lightbulbs!"
    print(p1+p2)

At Mercury the power is 25675 W which is equivalent to 430 lightbulbs!
At Venus   the power is 8384 W which is equivalent to 140 lightbulbs!
At Earth   the power is 4108 W which is equivalent to 68 lightbulbs!
At Mars    the power is 1826 W which is equivalent to 30 lightbulbs!
At Jupiter the power is 152 W which is equivalent to 2.5 lightbulbs!
At Saturn  the power is 45 W which is equivalent to 0.74 lightbulbs!
At Neptune the power is 5 W which is equivalent to 0.076 lightbulbs!
At Uranus  the power is 11 W which is equivalent to 0.19 lightbulbs!
At Pluto   the power is 3 W which is equivalent to 0.044 lightbulbs!


Amazing! The eagled eyed among you will have noticed that Neptune and Uranus are in the wrong order in the lists I gave you. This hasn't been a problem up to now but assume you want to get the power for Neptune from the ```Pmro``` list. It should be the 8th element with index 7 right? But I swapped the planets so it isn't and we shouldn't assume so.

Yikes, there must be some way to get the index of Neptune from the ```names``` array and use that to access the correct value in ```Pmro```.
<a class="anchor" id="WHERE"></a>
This is where the NumPy function ```np.where()``` comes in handy. We can use it like this:

```Python
cheese = np.asarray(['normal', 'smelly', 'holey'])
ind = np.where(cheese == 'holey')
```

This will tell us where the value ```'holey'``` is in the ```cheese``` array.

Use ```np.where()``` to find the index where ```names == 'Neptune'```. Print out the variable ```ind```.

In [14]:
ind = np.where(names == 'Neptune')
ind

(array([6], dtype=int64),)

Here you should notice that it didn't return a number but a tuple. The first entry in the tuple is our array of index values where NumPy found the string Neptune. In the cell below print the first element of the tuple.

In [15]:
ind[0]

array([6], dtype=int64)

In the next cell print the first value of this array.

In [16]:
ind[0][0]

6

So we have to type ```ind[0][0]``` to get the index where Neptune was found. Why? Because with higher dimensional data the tuple contains multiple arrays and if Neptune appeared more than once then the array would contain more than one value.

Use this index to access the correct value in ```Pmro_ratio``` and print it below:

In [17]:
Pmro_ratio[ind[0][0]]

0.07607485532219052

## Turning it into a Function

Let's try and turn this into a function. Complete the code below to finish the ```solar_panel_power()``` function.

* complete the ```ind``` line using ```np.where()``` - remember the name of the planet is being supplied by the user as the function argument
* add in the ```p1``` and ```p2``` lines from above
* test the function by running ```solar_panel_power('Mars')``` and checking the output is the same as before

In [18]:
def solar_panel_power(name):
    # takes the planet name - prints the power of the solar panels at the planet
    ind = np.where(names == name)[0][0]
    p1 = f'At {name:<7} the power is {Pmro[ind]:.0f} W which is equivalent to '
    p2 = f"{float(f'{Pmro_ratio[ind]:.2g}'):g} lightbulbs!"
    print(p1+p2)
    
solar_panel_power('Mars')

At Mars    the power is 1826 W which is equivalent to 30 lightbulbs!


Ok great but this function relies on the constants we defined above and the power arrays which were calculated outside the function. In the next code cell I have moved the constants and calculations inside the function.

Is this more or less readable than before? Do you think this is good practice?

In [19]:
def solar_panel_power2(name):
    # takes the planet name - prints the power of the solar panels at the planet
    sigma = 5.670373*10**(-8) # Wm^-2K^-4
    Rsun = 696340000 # m
    Tsun = 5778 # K
    AU = 1.495978707*10**11 # m
    Lsun = 4*np.pi*Rsun**2*sigma*Tsun**4
    names = ['Mercury', 'Venus', 'Earth', 'Mars', 'Jupiter', 'Saturn', 'Neptune', 'Uranus', 'Pluto']
    distances = [0.4, 0.7, 1.0, 1.5, 5.2, 9.6, 30, 19.2, 39.5]
    names = np.asarray(names)
    distances = np.asarray(distances)
    b = Lsun/(4*np.pi*(distances*AU)**2) # W/m^2
    area = 10 # m^2
    eff = .3
    Pmro = b*area*eff
    bulb = 60 # W
    Pmro_ratio = Pmro/bulb
    ind = np.where(names == name)[0][0]
    p1 = f'At {name:<7} the power is {Pmro[ind]:.0f} W which is equivalent to '
    p2 = f"{float(f'{Pmro_ratio[ind]:.2g}'):g} lightbulbs!"
    print(p1+p2)
    
solar_panel_power2('Mars')

At Mars    the power is 1826 W which is equivalent to 30 lightbulbs!


This may be tempting but it is not easy to read! Plus we just need the value for Mars now so there is no point calculating powers for all the planets.

The first step to tidy this up is to create a planets dictionary. We define this before defining the function and it contains the planet names as keys and the distances as values. It has been started for you. 

* Complete the dictionary using data from the lists at the start
* Change the ```distances``` line to ```distance``` and select the correct distance from the dictionary

Remember if you want the value from the dictionary you type ```dictionary[key]```.

In [20]:
planets = {
    'Mercury': 0.4,
    'Venus': 0.7,
    'Earth': 1.0,
    'Mars': 1.5,
    'Jupiter': 5.2,
    'Saturn': 9.6,
    'Neptune': 30,
    'Uranus': 19.2,
    'Pluto': 39.5
}

def solar_panel_power3(name):
    # takes the planet name - prints the power of the solar panels at the planet
    sigma = 5.670373*10**(-8) # Wm^-2K^-4
    Rsun = 696340000 # m
    Tsun = 5778 # K
    AU = 1.495978707*10**11 # m
    Lsun = 4*np.pi*Rsun**2*sigma*Tsun**4
    distance = planets[name]
    b = Lsun/(4*np.pi*(distance*AU)**2) # W/m^2
    area = 10 # m^2
    eff = .3
    Pmro = b*area*eff
    bulb = 60 # W
    Pmro_ratio = Pmro/bulb
    p1 = f'At {name:<7} the power is {Pmro:.0f} W which is equivalent to '
    p2 = f"{float(f'{Pmro_ratio:.2g}'):g} lightbulbs!"
    print(p1+p2)
    
solar_panel_power3('Mars')

At Mars    the power is 1826 W which is equivalent to 30 lightbulbs!


Ok now we can look at potentially removing something else from the function.
The constants we defined can come out of the function and be defined all together. 
This has the added benefit that the constants can be used by other functions not just the one we are writing now.

Move the constants out of the function into the space provided for you.

In [21]:
planets = {
    'Mercury': 0.4,
    'Venus': 0.7,
    'Earth': 1.0,
    'Mars': 1.5,
    'Jupiter': 5.2,
    'Saturn': 9.6,
    'Neptune': 30,
    'Uranus': 19.2,
    'Pluto': 39.5
}

##### Define Constants
sigma = 5.670373*10**(-8) # Wm^-2K^-4
AU = 1.495978707*10**11 # m
bulb = 60 # W

Rsun = 696340000 # m
Tsun = 5778 # K
#####

def solar_panel_power4(name):
    # takes the planet name - prints the power of the solar panels at the planet
    Lsun = 4*np.pi*Rsun**2*sigma*Tsun**4
    distance = planets[name]
    b = Lsun/(4*np.pi*(distance*AU)**2) # W/m^2
    area = 10 # m^2
    eff = .3
    Pmro = b*area*eff
    Pmro_ratio = Pmro/bulb
    p1 = f'At {name:<7} the power is {Pmro:.0f} W which is equivalent to '
    p2 = f"{float(f'{Pmro_ratio:.2g}'):g} lightbulbs!"
    print(p1+p2)
    
solar_panel_power4('Mars')

At Mars    the power is 1826 W which is equivalent to 30 lightbulbs!


This is more like how you would structure a function and the data it requires in a non-interactive Python file (a normal **.py** file). It is far more readable than chucking everything the function needs inside it.

## Function Arguments <a class="anchor" id="FA"></a>

The last thing we will do is look at the [arguments](https://www.programiz.com/python-programming/function-argument) of our function.

The arguments of the function are anything that goes in the ```()``` when defining a function (and using it). Here ```name``` is our only argument so far. If an argument is defined like ```name``` has been then it is a standard argument.

Notice we assumed ```name``` is a string and one of the keys in our ```planets``` dictionary. For now this is fine but if we were going to publish this code the function docstring (the comment after we start defining the function) would need to be altered to tell the user that the ```name``` argument must be a string and a key of the planets dictionary. After we make our changes to the arguments of our function we will update our docstring to let users know what types they must be.

There are three types of arguments we can supply to our function:

* Standard arguments
* \*args arguments
* \*\*kwargs arguments 

They must be supplied in the order listed above.

**Standard arguments** are defined as shown below and can have default values:

```Python
def carrots(num, colour='Orange'):
    print(str(num)+colour)
    
carrots(5)
carrots(5, 'purple')
```
Output:
```
>>> 5 Orange
>>> 5 purple
```
Here the standard argument ```colour``` has been given the default value ```'Orange'```. This means if the user doesn't pass the colour argument to the function it defaults to ```'Orange'```.

**\*args** lets us pass a varying number of arguments to the function. With this we can pass any number of arguments we like as in the example below:

```Python
def print_arguments(*args):
    # prints the arguments passed to this function
    for i in *args:
        print(i)
    return   

print_arguments('hello')
print_arguments('size', 'mass') 
```
Output:
```
>>> 'hello'
>>> 'size'
>>> 'mass'
```

If I changed the number of arguments passed to ```print_arguments()``` above the function still runs since it can take any number of arguments! You don't have to call these arguments ```*args``` - the asterik is the magic part. It's an unpacking operator which gives you a tuple of arguments to 'unpack' and use. I could have called ```*args``` ```*my_args``` or anything else and it would still work.

**\*\*kwargs arguments** or keyword arguments are added with ```**kwargs``` like in the following example:

```Python
def calculate(number, **kwargs):
    power = kwargs.get('area', None)
    if power:
        return number**power
    return number
    
calculate(7, power=2)
```

This code snippet lets the user specify a ```power``` to raise their number by. The keyword arguments are passed to the function as a dictionary so we get the keyword argument using the ```kwargs.get(KEYWORD, DEFAULT)``` function. Notice here if the user didn't specify a power then it would default to a value of ```None``` and the number would have been returned unchanged. Again we don't have to call the dictionary ```kwargs``` the double asteric unpacking operator is the important part.

For more information on arguments and the unpacking operators, ```*``` and ```**```, see [this article](https://realpython.com/python-kwargs-and-args/#unpacking-with-the-asterisk-operators) on realpython.com. 

We are going to change our function so that the ```area``` and ```eff``` of the solar panels can be changed by the user. If they are not specified they can default to the values we were using before.

* Add ```**kwargs``` to the function definition
* Change the ```area``` and ```eff``` variables to get the values from the keyword arguments dictionary using ```.get()``` - make sure you set the defaults to the correct numbers
* Call the function with ```solar_panel_power5('Mars', area=10, eff=.3)``` to check if it works

In [22]:
planets = {
    'Mercury': 0.4,
    'Venus': 0.7,
    'Earth': 1.0,
    'Mars': 1.5,
    'Jupiter': 5.2,
    'Saturn': 9.6,
    'Neptune': 30,
    'Uranus': 19.2,
    'Pluto': 39.5
}

##### Define Constants
sigma = 5.670373*10**(-8) # Wm^-2K^-4
AU = 1.495978707*10**11 # m
bulb = 60 # W

Rsun = 696340000 # m
Tsun = 5778 # K
#####

def solar_panel_power5(name, **kwargs):
    # takes the planet name - prints the power of the solar panels at the planet
    Lsun = 4*np.pi*Rsun**2*sigma*Tsun**4
    distance = planets[name]
    b = Lsun/(4*np.pi*(distance*AU)**2) # W/m^2
    area = kwargs.get('area', 10)
    eff = kwargs.get('eff', .3)
    Pmro = b*area*efficiency
    Pmro_ratio = Pmro/bulb
    p1 = f'At {name:<7} the power is {Pmro:.0f} W which is equivalent to '
    p2 = f"{float(f'{Pmro_ratio:.2g}'):g} lightbulbs!"
    print(p1+p2)
    
solar_panel_power5('Mars', area=10, eff=.3)

At Mars    the power is 1826 W which is equivalent to 30 lightbulbs!


We can now change the area and efficiency of the solar panels to see how they affect the power supplied.

The last thing we are going to add is the option to change what sort of star we are orbiting. We can do this by adding a new keyword argument called ```star="Sun"```. Notice I've put the keyword equal to a string. This means our argument will default to ```"Sun"``` if it isn't specified when the function is called by the user.

* Add ```star="Sun"``` after the ```name``` argument
* Use an if statement to check if ```star == "Sun"```. If it is use ```Tsun``` and ```Rsun``` in your calculation for ```Psun```
* else assume a tuple was passed to star like this: ```star=(radius, temp)```. Unpack the tuple and use the supplied radius and temperature to calculate ```Psun```
* Check your code by running ```solar_panel_power6('Mars', star=(800000000, 7000), area=10, eff=.3)```

In [23]:
planets = {
    'Mercury': 0.4,
    'Venus': 0.7,
    'Earth': 1.0,
    'Mars': 1.5,
    'Jupiter': 5.2,
    'Saturn': 9.6,
    'Neptune': 30,
    'Uranus': 19.2,
    'Pluto': 39.5
}

##### Define Constants
sigma = 5.670373*10**(-8) # Wm^-2K^-4
AU = 1.495978707*10**11 # m
bulb = 60 # W

Rsun = 696340000 # m
Tsun = 5778 # K
#####

def solar_panel_power6(name, star="Sun", **kwargs):
    """Calculates the power from the solar panels at the planet
    
    Takes the name of the planet and details about the star
    outputs the power that the solar panels would deliver
    
    Args:
        name -- string, must be in the planets dictionary
        star -- tuple, (RADIUS, TEMP) of star in m and K (default to Sun's values)
        area -- float, area of solar panels in m^2 (default 10)
        eff -- float, efficiency of the solar panels (default .3)
        
    Returns:
        Nothing - prints the values to the terminal
        
    Example:
        >>> solar_panel_power6('Mars', star=(800000000, 7000), area=10, eff=.3)
        At Mars the power is 5191 W which is equivalent to 87 lightbulbs!
    """
    if star == "Sun":
        R, T = Rsun, Tsun
    else:
        R, T = star[0], star[1]
    Lsun = 4*np.pi*R**2*sigma*T**4
    distance = planets[name]
    b = Lsun/(4*np.pi*(distance*AU)**2) # W/m^2
    area = kwargs.get('area', 10)
    eff = kwargs.get('eff', .3)
    Pmro = b*area*efficiency
    Pmro_ratio = Pmro/bulb
    p1 = f'At {name:<7} the power is {Pmro:.0f} W which is equivalent to '
    p2 = f"{float(f'{Pmro_ratio:.2g}'):g} lightbulbs!"
    print(p1+p2)
    
solar_panel_power6('Mars', star=(800000000, 7000), area=10, eff=.3)

At Mars    the power is 5191 W which is equivalent to 87 lightbulbs!


<a class="anchor" id="DS"></a>
Now it's time to update the docstring. This isn't a simple function so we are probably going to need a multi-line docstring. Ideally it will be formatted like this:

```Python
def cheese_burger(quant, lettuce=True, **kwargs):
    """Outputs ingredient quantities for cheeseburger
    
    Creates an ingredients.txt file with ingredient quantities based off user input
    
    Args:
        quant -- int, the number of burgers (must be non zero and positive)
        lettuce -- boolean, add lettuce or not (default True)
        tomatoe -- boolean, add tomatoe or not (default False)
    
    Returns:
        Nothing - prints quantities to file
        
    Raises:
        TypeError: If quant is not an integer above 0
        
    Example:
        >>> cheese_burger(4, lettuce=True)
        File ingredients.txt was created
    """
```

This is one of MANY ways of formatting a docstring but I find it to be the most readable. You can see some other examples on this [stackoverflow post](https://stackoverflow.com/questions/3898572/what-is-the-standard-python-docstring-format).

The first line is a brief summary.    
Then you can write an extended summary if needed (don't let text span over 70 characters - use multiple lines!).    
Then define the arguments.     
State what the function returns (if anything).     
State any errors that the function raises (ours doesn't for now).    
Finally you can add some example usage cases.    

Go ahead and update the function's docstring. Then if you're up for more tackle some of the challenges below.

## Over to You <a class="anchor" id="OTY"></a>

Have a go at the tasks below. I haven't provided solutions and there are probably a few ways to solve each one! Remember if you are stuck you can consult the Python and NumPy docs at https://docs.python.org/3/, and https://numpy.org/doc/stable/, find help from one of these websites www.w3schools.com, www.python.org, www.learnpython.org or you can also get advice from www.stackoverflow.com. You can also ask your instructor for help or email me at astrodimitrios@gmail.com.

**Task 1:** Alter your function to return the sentence and store it in a variable.    
**Task 2:** If you want to run the code with a fictional planet at a different distance we can have the user pass a name AND a distance. Change the name argument to planet. In your function check if it is a string. If it is then get the distance from the dictionary. If it isn't assume a tuple was passed in the form (name, distance). This is similar to what we did with the star keyword argument.     
**Task 3:** Assume that the satellite only works with a power greater than 1000 W - have the function also calculate the maximum distance away from the host star the satellite could function at.    
**Task 4:** Test the area keyword argument before using it to see if it is a float (a number and not a string or a list etc.) and greater than zero, if it isn't raise an exception, see [here](https://www.w3schools.com/python/python_try_except.asp). You may need to use another if statement.    
**Task 5:** You may remember from last lesson if a dictionary is only being used by one function we should try and define it inside the function or pass it as an argument to the function. Alter your function to accept the ```planets``` dictionary as an argument.

In [24]:
# Your code here:


## Sharing

If you share, use or modify this activity in any way use the citation in this [txt file](https://github.com/astroDimitrios/Astronomy/blob/master/CITATION.txt).    
Please contact me at astrodimitrios@gmail.com with any suggestions, mistakes found, or general questions about teaching astronomy with Python.

© Dimitrios Theodorakis GNU General Public License v3.0 https://github.com/astroDimitrios/Astronomy  